File: Sliding_Window/longest-substring-without-repeating-characters.go
package main

/*
Given a string s, find the length of the longest substring without repeating characters.



Example 1:

Input: s = "abcabcbb"
Output: 3
Explanation: The answer is "abc", with the length of 3.
Example 2:

Input: s = "bbbbb"
Output: 1
Explanation: The answer is "b", with the length of 1.
Example 3:

Input: s = "pwwkew"
Output: 3
Explanation: The answer is "wke", with the length of 3.
Notice that the answer must be a substring, "pwke" is a subsequence and not a substring.


Constraints:

0 <= s.length <= 5 * 104
s consists of English letters, digits, symbols and spaces.
*/
// Approach 1 Brute force
// Intuition
// Check all the substring one by one to see if it has no duplicate character.
// Time complexity O(n³ ) Space complexity O(m) where m is size of hMap
func lengthOfLongestSubstring(s string) int {
    n := len(s)
    res := 0
    for i := 0; i < n; i++ {
        for j := i; j < n; j++ {
            if check(s, i, j) {
                res = max(res, j - i + 1)
            }
        }
    }
    return res
}
func max(a, b int) int {
    if a >= b {
        return a
    } else {
        return b
    }
}
func check(s string, start int, end int) bool {
    hMap := [128]int{}
    for i := start; i <= end; i++ {
        hMap[s[i]]++
        if hMap[s[i]] > 1 {
            return false
        }
    }
    return true
}

File: Strings/longest_substring_without_repeating_characters.cpp
/**
  Time Complexity: t(n) = n * 128 -> O(n).
  Where n is the size of string s.

  Space Complexity: O(1)

  Given a string s, find the length of the longest substring without repeating characters.

  Example 1:
  Input: s = "abcabcbb"
  Output: 3
  Explanation: The answer is "abc", with the length of 3.

  Example 2:
  Input: s = "bbbbb"
  Output: 1
  Explanation: The answer is "b", with the length of 1.

  Example 3:
  Input: s = "pwwkew"
  Output: 3
  Explanation: The answer is "wke", with the length of 3.
  Notice that the answer must be a substring, "pwke" is a subsequence and not a substring.

  Constraints:
    0 <= s.length <= 5 * 104
    s consists of English letters, digits, symbols and spaces.

**/


class Solution {
public:



    // Array to store the frequency of characters for sub-string that's starts at index a and end at index b - 1.
    int v[128] = {0};

    /**
        Function: mxDistinct returns the maximum number of unique charachters of substring s[a, b[
        if there is at least one character that present more than one time the function returns 0,
        beacause where are looking for the lonest substring withput repating characters.
    **/
    int mxDistinct()
    {
        int c = 0;
        for (int i = 0; i < 128; ++i) {
            if(v[i] > 1)
                return 0;

            if(v[i] == 1)
                c++;
        }
        return c;
    }

    int lengthOfLongestSubstring(string s) {

        int a = 0, b = 0;
        int sz = s.size();

        if(!sz) // If the string s is empty the functon returns 0.
            return 0;

        int mx = 0; // Variable mx to store the result.

        while (a <= b && b < sz) 
        {

            v[s[b++] - ' ']++; // Count how many times character at index b (s[b]) appears in the string s.

            int curMax = mxDistinct(); // Variable to store the current max or the max so far.
	    if(curMax) /**
                           If we find a substring without repeating characters we take the maximum between mx and curMax.
	                    REMEMBER: We are looking for the longest one. Actullay, the size of the longest one.
	               **/

                mx = max(mx, curMax);
            else // If there is no such a substring we start looking for substring that begins at index a + 1.
                v[s[a++] - ' ']--;
        }

        return mx;
    }
};
